clusterscope/cluster_info.py

```
@@ -7,12 +7,114 @@
+
+        Returns:
+            int: The number of CPUs per node, assuming all nodes have the same CPU count.
+        """
+        if self.slurm_cluster_info.is_slurm_cluster:
+            return self.slurm_cluster_info.get_cpus_per_node()
+        return self.local_node_info.get_cpu_count()
+
+    def get_gpu_generation_and_count(self) -> Dict[str, int]:
+        """Get the number of GPUs on the slurm cluster node.
+
+        Returns:
+            dict: A dictionary with GPU generation as keys and counts as values.
+        """
+        if self.slurm_cluster_info.is_slurm_cluster:
+            return self.slurm_cluster_info.get_gpu_generation_and_count()
+        return self.local_node_info.get_gpu_generation_and_count()
+
+
+class LocalNodeInfo:
+    """A class to provide information about the local node.
+
+    This class offers methods to query various aspects of the local node,
+    such as CPU and GPU information.
+    """
+
+    @fs_cache(var_name="LOCAL_NODE_CPU_COUNT")
+    def get_cpu_count(self, timeout: int = 60) -> int:
+        """Get the number of CPUs on the local node.
+
+        Returns:
+            int: The number of CPUs on the local node.
+
+        Raises:
+            RuntimeError: If unable to retrieve CPU information.
+        """
+        try:
+            result = subprocess.check_output(
+                ["nproc", "--all"], text=True, timeout=timeout
+            )
+
+            return int(result.strip())
+        except (subprocess.SubprocessError, FileNotFoundError) as e:
+            raise RuntimeError(f"Failed to get CPU information: {str(e)}")
+
+    def get_gpu_generation_and_count(self, timeout: int = 60) -> Dict[str, int]:
+        """Get the number of GPUs on the local node.
+
+        Returns:
+            int: The number of GPUs on the local node.
+
+        Raises:
+            RuntimeError: If unable to retrieve GPU information.
+        """
```

Why should we throw an exception?

Imagine someone running this on a submit node in one of our clusters, shouldn't we return 0 GPUs instead of failing the code?

We need to distinguish between nvidia-smi failing to run Vs. succeeding, but, no GPUs present on a node.

---

yeah, good idea. now we're distinguishing

---

Sorry, not clear yet.

There are three error possibilities as I see it:

1. nvidia-smi command is available, user runs on a GPU-ful node and the command errors out
2. nvidia-smi is available, runs on a submit node
3. nvidia-smi is NOT available

In case-1, exception makes sense as we couldn't retrieve the gpu information, either transiently or permanently.

For case-2, what do we want to see? 0 is a local answer, but, not  what the user is looking for as they plan to run the code in the cluster.

For case-3, ideally we return 0.

---

> 3. nvidia-smi is NOT available

we're checking this with `has_nvidia_gpus`, added caching and assert on this function